### PR TITLE
Update Java client to Search API v3

### LINKED ISSUE #28: HTTP compression
We now have support for HTTP compression (e.g. gzip) for Twingly Blog Search API, but this client does not ask for compression.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/42020228-http-compression?utm_campaign=plugin&utm_content=tracker%2F3240456&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F3240456&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #29: Update Java client to Search API v3 [$50]
Update the client to use the new [Blog Search API v3](https://developer.twingly.com/resources/search/).

Notable changes from the previous version of the Blog Search API are listed under https://developer.twingly.com/resources/search#changes-from-previous-version.

The ruby client was updated in https://github.com/twingly/twingly-search-api-ruby/pull/65, which can be used as help. Some XML fixtures can also be copied from there if needed.

If possible, try to keep all classes/methods/properties backwards compatible with the previous version of the client by marking methods as deprecated. Similar to what was done in the Ruby client.

<bountysource-plugin>

---
There is a **[$50 open bounty](https://www.bountysource.com/issues/45000489-update-java-client-to-search-api-v3?utm_campaign=plugin&utm_content=tracker%2F3240456&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F3240456&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by xSAVIKx**
I've updated the Java client to the latest v3 API.

What was done:
* dependencies versions were bumped to the latest ones
* gradle version was bumped to the latest one
* added .gitattribute in order to be sure that files have correct line endings on commits
* updated the overall client classes to work with v3 API - can close #29.
* added support for compression - can close #28.

### CODE REVIEW

**Comment by dentarg** [comment]
Hi @xSAVIKx 

This is not a proper review from me, I have just some questions. @roback will review, hopefully later this week.

> disable openJDK build as it fails with sigfaults

Care to elaborate on this? Do you know the reason to as it sigfaults (did you mean "segfault"?)? Does the old version have the same problem with openjdk7?

Do we really need a `.gitattributes` that long for this repo?

Thanks!

File: README.md [hunk 1]
```diff
@@ -151,13 +151,13 @@ This will save JAR file with all needed dependencies in `build/libs/` with `-all
...
-1. Log into [Nexus Repository Manager](https://oss.sonatype.org/) and find the archive
-1. Read [Releasing the Deployment](http://central.sonatype.org/pages/releasing-the-deployment.html)
-1. Release to Central Repository using [Nexus Repository Manager](https://oss.sonatype.org/)
+2. Bump the version in [version.properties](./src/main/resources/version.properties)
+3. Commit the changes
+4. Tag the current commit with the same version number and push it
+5. Run `gradlew uploadArchives -Prelease` to release to the staging repository
+6. Log into [Nexus Repository Manager](https://oss.sonatype.org/) and find the archive
+7. Read [Releasing the Deployment](http://central.sonatype.org/pages/releasing-the-deployment.html)
+8. Release to Central Repository using [Nexus Repository Manager](https://oss.sonatype.org/)
```
Comment (dentarg) [inline README.md]: Actually, when rendered, Markdown will do the numbering, not having the order in the Markdown source makes for easier "refactoring" :-)

From https://guides.github.com/features/mastering-markdown/

Source:

```markdown
1. Item 1
1. Item 2
1. Item 3
   1. Item 3a
   1. Item 3b
```

Rendered:

1. Item 1
1. Item 2
1. Item 3
   1. Item 3a
   1. Item 3b

Comment (xSAVIKx) [inline README.md]: Yeah, but I mostly read raw markdown, that's why I'm always trying to use correct numbering.

If you do think it should be reverted to serie of `1.` - I'd do that.

**Comment by xSAVIKx** [comment]
@dentarg Yes, misprint - segfault.
I do not know the reason and actually not willing to dig into the OpenJDK. It was introduced just as an option and I don't think anyone should invest time into making it work on Travis OpenJDK.

As for the `.gitattributes` - it's just a generated boilerplate for any possible project. If you do want - I can clean it up.

File: README.md (hunk 1 above)
Comment (dentarg) [inline README.md]: Yes, please revert to `1.`s

**Comment by dentarg** [comment]
Yes, please cleanup `.gitattributes`

**Comment by xSAVIKx** [comment]
@roback @dentarg Please review updates.

I've made client more backward-compatible - almost same changes as for the Python client.

File: src/main/java/com/twingly/search/Constants.java [hunk 2]
```diff
@@ -12,7 +12,12 @@
     /**
      * The constant DATE_FORMAT.
      */
-    public static final String DATE_FORMAT = "yyyy-MM-dd HH:mm:ss";
+    public static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ssX";
```
Comment (xSAVIKx) [inline src/main/java/com/twingly/search/Constants.java]: @roback nope.

`Z` in Java is for RFC standard only. And `X` is for ISO standard.

File: src/main/java/com/twingly/search/Query.java [hunk 3]
```diff
@@ -1,94 +1,228 @@
...
+
+    /**
+     * Convert Query object to its String representation
+     *
+     * @return search query in String representation
+     */
+    public String toStringRepresentation() {
+        StringBuilder result = new StringBuilder(searchQuery);
+        if (lang != null) {
+            result.append(" lang: ").append(lang);
```
Comment (xSAVIKx) [inline src/main/java/com/twingly/search/Query.java]: OK

File: README.md [hunk 4]
```diff
@@ -15,9 +15,8 @@ public class Test{
         String apiKey = "API_KEY";
         // create Client object using your API key
         Client client = new UrlConnectionClient(apiKey);
-        // Create Query object using QueryBuilder for results in English published from startTime to endTime
-        Query query = QueryBuilder.create("spotify").documentLanguage(Language.English).startTime(startTime).endTime(endTime).build();
-        Result result = client.makeRequest(query);
+        Query q = "spotify";
```
Comment (xSAVIKx) [inline README.md]: Yeah :-) Or realy `Query` object.

File: src/main/java/com/twingly/search/Query.java [hunk 5]
```diff
@@ -1,94 +1,228 @@
...
+        this.location = location;
+    }
+
+    /**
+     * Gets language
+     *
+     * @return language
+     * @since 3.0.0
+     */
+    public String getLang() {
```
Comment (roback) [inline src/main/java/com/twingly/search/Query.java]: There is no need for `getLang` and `setLang`. It's enough with deprecating `get/setDocumentLanguage` (as you have already done below).

File: src/main/java/com/twingly/search/Query.java [hunk 6]
```diff
@@ -1,94 +1,228 @@
...
+        return QueryBuilder.fromQuery(this);
+    }
+
+    /**
+     * Gets location
+     *
+     * @return the location
+     * @since 3.0.0
+     */
+    public String getLocation() {
```
Comment (roback) [inline src/main/java/com/twingly/search/Query.java]: Remove `get/setLocation` (they aren't used anywhere). Location is part of the query string.

File: src/main/java/com/twingly/search/Query.java (hunk 5 above)
Comment (roback) [inline src/main/java/com/twingly/search/Query.java]: Ignore my comment :) You can go ahead with `get/setLang`. I got a bit confused as the Java client uses a query builder which none of the other clients does.

File: src/main/java/com/twingly/search/Query.java (hunk 6 above)
Comment (roback) [inline src/main/java/com/twingly/search/Query.java]: > A test that checks that the location gets added to the query string should be added as well.

Not just a test for location, but a test that all attributes gets added to the query correctly.

Comment (xSAVIKx) [inline src/main/java/com/twingly/search/Query.java]: @roback done.

File: src/main/java/com/twingly/search/Constants.java [hunk 7]
```diff
@@ -15,10 +15,6 @@
     public static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ssX";
 
     /**
-     * The constant DATE_FORMAT.
-     */
-    public static final String SEARCH_QUERY_DATE_FORMAT = "yyyy-MM-dd HH:mm:ss";
```
Comment (roback) [inline src/main/java/com/twingly/search/Constants.java]: Great! Thanks for taking the time to update this as well :)

**Comment by IuriiSergiichuk** [comment]
Hi @roback, how it is going?

**Comment by roback** [comment]
> Hi @roback, how it is going?

Unfortunately we have a few other things that has higher priority that we need to finish first, so I won't be able to continue with this at the moment. Sorry for keeping you waiting.

File: src/main/java/com/twingly/search/domain/Post.java [hunk 8]
```diff
@@ -300,4 +397,221 @@ public void setBlogRank(int blogRank) {
...
+    public void setImages(List<String> images) {
+        this.images = images;
+    }
+
+    /**
+     * Gets coordinates
+     *
+     * @return the coordinates
+     */
+    public String getCoordinates() {
```
Comment (roback) [inline src/main/java/com/twingly/search/domain/Post.java]: Posts containing coordinates doesn't work, it looks like it's always an empty string.

In the XML the coordinates looks like:

```xml
<coordinates>
  <latitude>49.1</latitude>
  <longitude>10.75</longitude>
</coordinates>
```

An example query for a post containing coordinates: https://api.twingly.com/blog/search/api/v3/search?apikey=KEY&q=id:7533026945919130930

In ruby we just put them in a hash: https://github.com/twingly/twingly-search-api-ruby/blob/dbf547f182d93137bb167eb99c9eda4a8197ac21/lib/twingly/search/parser.rb#L72-L75. I don't know what works best in Java though :)

File: example/src/main/java/SearchPostsStream.java [hunk 9]
```diff
@@ -44,8 +44,9 @@ private void iterateAllPosts(Result result) {
             return;
         }
         Post lastPost = result.getPosts().get(result.getPosts().size() - 1);
-        query.setStartTime(lastPost.getPublished());
-        Result newResult = client.makeRequest(query);
+        Date lastPublishedAt = lastPost.getPublishedAt();
+        q = q.toBuilder().startTime(lastPublishedAt).build();
+        Result newResult = client.makeRequest(q);
```
Comment (xSAVIKx) [inline example/src/main/java/SearchPostsStream.java]: @roback It's actually not the best idea to allow passing `QueryBuilder` into the client.

The reason for the `Query` object existence is that it is immutable. While `QueryBuilder` is mutable and we should try not pass mutable data to the client (as mutability could lead to unpredictable side effects).

I'd prefer to leave it as is. If you do insist - I'll change the implementation.

File: src/main/java/com/twingly/search/domain/Post.java (hunk 8 above)
Comment (xSAVIKx) [inline src/main/java/com/twingly/search/domain/Post.java]: @roback There was just no info about how coordinates are represented.
I'll add mapping for coordinate a bit later.

File: example/src/main/java/SearchPostsStream.java [hunk 10]
```diff
@@ -44,8 +44,9 @@ private void iterateAllPosts(Result result) {
             return;
         }
         Post lastPost = result.getPosts().get(result.getPosts().size() - 1);
-        query.setStartTime(lastPost.getPublished());
-        Result newResult = client.makeRequest(query);
+        Date lastPublishedAt = lastPost.getPublishedAt();
+        q = q.toBuilder().startTime(lastPublishedAt).build();
```
Comment (xSAVIKx) [inline example/src/main/java/SearchPostsStream.java]: Here we can use `QueryBuilder`, but we will have to call `build()` anyway.

File: example/src/main/java/SearchPostsStream.java (hunk 9 above)
Comment (roback) [inline example/src/main/java/SearchPostsStream.java]: :+1: Now that I know there is a reason for this we can leave it as is :)

**Comment by roback** [comment]
Tested the coordinates and it worked great 👍 

I will merge this now.

**Pull request merged**